src/main/java/org/radarcns/service/DataSetService.java

```diff
@@ -197,4 +241,66 @@ private Header getHeader(String project, String subject, String source,
         return new Header(project, subject, source, sourceType, sourceData.getSourceDataType(),
                 stat, sourceData.getUnit(), timeWindow, timeFrame, null);
     }
+
+    /**
+     * Returns calculated {@link AggregatedDataPoints} using given parameters.
+     *
+     * @param projectName of project
+     * @param subjectId of subject
+     * @param sources requested
+     * @param timeWindow interval
+     * @param start time
+     * @param end time
+     * @return calculated data.
+     */
+    public AggregatedDataPoints getAggregatedData(String projectName, String subjectId,
+            List<AggregateDataSource> sources, TimeWindow timeWindow, Instant start, Instant end) {
+        List<DataItem> dataItems = calculateIntervals(start, end, timeWindow).stream()
+                .map(p -> computeAggregatedDataItem(projectName, subjectId, sources, p, timeWindow))
+                .collect(Collectors.toList());
+        // fill up sourceData.type field
+        sources.forEach(d -> d.getSourceData().forEach(e -> {
```

Comment:
since you're modifying the internal data structure and throwing and exception, a nested for loop might be clearer here, e.g.

```java
try {
  for (AggregateSourceData source : sources) {
    for (SourceData sourceData : source.getSourceData()) {
         SourceData definition = sourceCatalog.getSourceData(sourceData.getName());
         sourceData.setType(definition.getSourceDataType());
    }
  }
} catch (...) {...}